applications/GeoMechanicsApplication/custom_constitutive/mohr_coulomb_with_tension_cutoff.cpp

```diff
@@ -0,0 +1,246 @@
+    return result;
+}
+
+Vector MohrCoulombWithTensionCutOff::ReturnStressAtCornerReturnZone(const Vector& rPrincipalTrialStressVector,
+                                                                    const Vector& rCornerPoint) const
+{
+    Vector result = rPrincipalTrialStressVector;
+    result(0)     = rCornerPoint(0) + rCornerPoint(1);
+    result(2)     = rCornerPoint(0) - rCornerPoint(1);
+    return result;
+}
+
+Vector MohrCoulombWithTensionCutOff::ReturnStressAtRegularFailureZone(const Vector& rPrincipalTrialStressVector,
+                                                                      const CoulombYieldSurface& rCoulombYieldSurface,
+                                                                      double FrictionAngle,
+                                                                      double Cohesion) const
+{
+    Vector result = rPrincipalTrialStressVector;
+    Vector flowDerivative = rCoulombYieldSurface.DerivateOfFlowFunction(rPrincipalTrialStressVector);
+    double cof1   = (1.0 + std::sin(FrictionAngle)) / (1.0 - std::sin(FrictionAngle));
+    double cof2   = 2.0 * Cohesion * std::cos(FrictionAngle) / (1.0 - std::sin(FrictionAngle));
+    double det    = cof1 * flowDerivative(0) - flowDerivative(2);
+    double lambda = rPrincipalTrialStressVector(2) + cof2 - rPrincipalTrialStressVector(0) * cof1;
+    lambda /= det;
+    result(0) = lambda * flowDerivative(0) + rPrincipalTrialStressVector(0);
+    result(2) = result(0) * cof1 - cof2;
+    return result;
+}
+
+double MohrCoulombWithTensionCutOff::CalculateApex(double FrictionAngle, double Cohesion) const
+{
+    return Cohesion / std::tan(FrictionAngle);
+}
+
+Vector MohrCoulombWithTensionCutOff::CalculateCornerPoint(double FrictionAngle, double Cohesion, double TensionCutoff) const
+{
+    Vector result = ZeroVector(2);
+    if (TensionCutoff < result(0)) {
+        result(0) = this->CalculateApex(FrictionAngle, Cohesion);
+        return result;
+    }
+    result(0) = (TensionCutoff - Cohesion * std::cos(FrictionAngle)) / (1.0 - std::sin(FrictionAngle));
+    result(1) = (Cohesion * std::cos(FrictionAngle) - TensionCutoff * std::sin(FrictionAngle)) /
+                (1.0 - std::sin(FrictionAngle));
+    return result;
+}
+
+Vector MohrCoulombWithTensionCutOff::RotatePrincipalStresses(const Vector& rPrincipalStressVector,
+                                                             const Matrix& rRotationMatrix) const
+{
+    Matrix principal_stress_matrix = GeoMechanicsMathUtilities::VectorToDiagonalMatrix(rPrincipalStressVector);
+    Vector result = StressStrainUtilities::RotateStressMatrix(principal_stress_matrix, rRotationMatrix);
+    return result;
+}
+
+void MohrCoulombWithTensionCutOff::CalculateTrialStressVector(const Vector& rStrainVector,
+                                                              Vector&       rStressVector,
+                                                              ConstitutiveLaw::Parameters& rValues) const
```

Comment:
This parameter is never used. That is very confusing. What about using this parameter and letting the caller provide it (e.g. by calling `rValues.GetStrainVector`)?

Also, I would suggest to use a proper return type here, rather than passing its result back through the parameter list:
```suggestion
Vector MohrCoulombWithTensionCutOff::CalculateTrialStressVector(const Vector& rStrainVector,
                                                                ConstitutiveLaw::Parameters& rValues) const
```

In addition, we could also supply Young's modulus and Poisson's ratio rather than the full `ConstitutiveLaw::Parameters`.